core/toolbox/category.js

```diff
@@ -0,0 +1,826 @@
+  var styleName = categoryDef['categorystyle'] || categoryDef['categoryStyle'];
+  var colour = categoryDef['colour'];
+
+  if (colour && styleName) {
+    console.warn('Toolbox category "' + this.name_ +
+        '" must not have both a style and a colour');
+  } else if (styleName) {
+    return this.getColourfromStyle_(styleName);
+  } else {
+    return this.parseColour_(colour);
+  }
+  return '';
+};
+
+/**
+ * Retrieves and sets the colour for the category using the style name.
+ * The category colour is set from the colour style attribute.
+ * @param {string} styleName Name of the style.
+ * @return {string} The hex colour for the category.
+ * @private
+ */
+Blockly.ToolboxCategory.prototype.getColourfromStyle_ = function(styleName) {
+  var theme = this.workspace_.getTheme();
+  if (styleName && theme) {
+    var style = theme.categoryStyles[styleName];
+    if (style && style.colour) {
+      return this.parseColour_(style.colour);
+    } else {
+      console.warn('Style "' + styleName +
+          '" must exist and contain a colour value');
+    }
+  }
+  return '';
+};
+
+/**
+ * Sets the colour on the category.
+ * @param {number|string} colourValue HSV hue value (0 to 360), #RRGGBB string,
+ *     or a message reference string pointing to one of those two values.
+ * @return {string} The hex colour for the category.
+ * @private
+ */
+Blockly.ToolboxCategory.prototype.parseColour_ = function(colourValue) {
+  // Decode the colour for any potential message references
+  // (eg. `%{BKY_MATH_HUE}`).
+  var colour = Blockly.utils.replaceMessageReferences(colourValue);
+  if (colour == null || colour === '') {
+    // No attribute. No colour.
+    return '';
+  } else {
+    var hue = Number(colour);
+    if (!isNaN(hue)) {
+      return Blockly.hueToHex(hue);
+    } else {
+      var hex = Blockly.utils.colour.parse(colour);
+      if (hex) {
+        return hex;
+      } else {
+        console.warn('Toolbox category "' + this.name_ +
+            '" has unrecognized colour attribute: ' + colour);
+        return '';
+      }
+    }
+  }
+};
+
+/**
+ * Opens or closes the current category if it has children.
+ * @param {boolean} isExpanded True to expand the category, false to close.
+ * @public
+ */
+Blockly.ToolboxCategory.prototype.setExpanded = function(isExpanded) {
+  if (!this.hasChildren() || this.expanded_ == isExpanded) {
+    return;
+  }
+  this.expanded_ = isExpanded;
+  if (isExpanded) {
+    this.subcategoriesDiv_.style.display = 'block';
+    this.openIcon_(this.iconDom_);
+  } else {
+    this.subcategoriesDiv_.style.display = 'none';
+    this.closeIcon_(this.iconDom_);
+  }
+  Blockly.utils.aria.setState(/** @type {!Element} */ (this.htmlDiv_),
+      Blockly.utils.aria.State.EXPANDED, isExpanded);
+
+  if (this.hasChildren()) {
+    for (var i = 0; i < this.getChildToolboxItems().length; i++) {
+      var child = this.getChildToolboxItems()[i];
+      child.isParentExpanded_ = isExpanded;
```

Comment:
You're directly accessing a private member of the child here, which is suspicious.

`isParentExpanded` is only about the direct parent, right? That is, no grandchildren would be marked by this.

Reply:
Fix in #4235